File: src/net/dinkla/raytracer/lights/AreaLight.java
package net.dinkla.raytracer.lights;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.hits.ShadowHit;
import net.dinkla.raytracer.materials.Material;
import net.dinkla.raytracer.math.*;
import net.dinkla.raytracer.objects.GeometricObject;
import net.dinkla.raytracer.worlds.World;

import java.util.ArrayList;
import java.util.List;


public class AreaLight extends Light implements ILightSource {

    public ILightSource object;

    // Emissive Material TODO: Warum nicht Emissive?
    public Material material;

    public int numSamples;

    public class Sample {
        public Point3D samplePoint;
        public Normal lightNormal;
        public Vector3D wi;

        public double getNDotD() {
            return lightNormal.negate().dot(wi);
        }
    }

    public AreaLight() {
        super();
        numSamples = 4;
    }

    public Color L(World world, Shade sr, Sample sample) {
        if (sample.getNDotD() > 0) {
            return sr.getMaterial().getLe(sr);
        } else {
            return Color.BLACK;
        }
    }

    public boolean inShadow(World world, Ray ray, Shade sr, Sample sample) {
        double d = sample.samplePoint.minus(ray.getO()).dot(ray.getD());
        return world.inShadow(ray, sr, d);
    }

    public double G(Shade sr, Sample sample) {
        double nDotD = sample.getNDotD();
        double d2 = sample.samplePoint.distanceSquared(sr.getHitPoint());
        return nDotD / d2;
    }

    public double pdf(Shade sr) {
        return object.pdf(sr);
    }

    public Sample getSample(Shade sr) {
        Sample sample = new Sample();
        sample.samplePoint = object.sample();
        sample.lightNormal = object.getNormal(sample.samplePoint);
        sample.wi = sample.samplePoint.minus(sr.getHitPoint()).normalize();
        return sample;
    }

    public List<Sample> getSamples(Shade sr) {
        List<Sample> result = new ArrayList<Sample>();
        for (int i = 0; i < numSamples; i++) {
            result.add(getSample(sr));
        }
        return result;
    }

    public Point3D sample() {
        throw new RuntimeException("NLU");
    }


    public Normal getNormal(Point3D p) {
        throw new RuntimeException("NLU");
    }

    @Override
    public Color L(World world, Shade sr) {
        throw new RuntimeException("NLU");
    }

    @Override
    public Vector3D getDirection(Shade sr) {
        throw new RuntimeException("NLU");
    }

    @Override
    public boolean inShadow(World world, Ray ray, Shade sr) {
        throw new RuntimeException("NLU");
    }

}

File: src/main/kotlin/net/dinkla/raytracer/lights/AreaLight.kt
package net.dinkla.raytracer.lights

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.materials.IMaterial
import net.dinkla.raytracer.math.Normal
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.math.Ray
import net.dinkla.raytracer.math.Vector3D
import net.dinkla.raytracer.worlds.World
import java.util.*


class AreaLight : Light(), ILightSource {

    var `object`: ILightSource? = null

    // Emissive Material TODO: Warum nicht Emissive?
    var material: IMaterial? = null

    var numSamples: Int = 0

    inner class Sample {
        var samplePoint: Point3D? = null
        var lightNormal: Normal? = null
        var wi: Vector3D? = null

        val nDotD: Double
            get() = (-lightNormal!!) dot (wi!!)
    }

    init {
        numSamples = 4
    }

    fun L(world: World, sr: Shade, sample: Sample): Color {
        return if (sample.nDotD > 0) {
            sr.material?.getLe(sr) ?: world.backgroundColor
        } else {
            Color.BLACK
        }
    }

    fun inShadow(world: World, ray: Ray, sr: Shade, sample: Sample): Boolean {
        val d = sample.samplePoint!!.minus(ray.origin).dot(ray.direction)
        return world.inShadow(ray, sr, d)
    }

    fun G(sr: Shade, sample: Sample): Double {
        val nDotD = sample.nDotD
        val d2 = sample.samplePoint!!.sqrDistance(sr.hitPoint)
        return nDotD / d2
    }

    override fun pdf(sr: Shade): Double {
        return `object`!!.pdf(sr)
    }

    fun getSample(sr: Shade): Sample {
        val sample = Sample()
        sample.samplePoint = `object`!!.sample()
        sample.lightNormal = `object`!!.getNormal(sample.samplePoint!!)
        sample.wi = sample.samplePoint!!.minus(sr.hitPoint).normalize()
        return sample
    }

    fun getSamples(sr: Shade): List<Sample> {
        val result = ArrayList<Sample>()
        for (i in 0 until numSamples) {
            result.add(getSample(sr))
        }
        return result
    }

    override fun sample(): Point3D {
        throw RuntimeException("NLU")
    }

    override fun getNormal(p: Point3D): Normal {
        throw RuntimeException("NLU")
    }

    override fun L(world: World, sr: Shade): Color {
        throw RuntimeException("NLU")
    }

    override fun getDirection(sr: Shade): Vector3D {
        throw RuntimeException("NLU")
    }

    override fun inShadow(world: World, ray: Ray, sr: Shade): Boolean {
        throw RuntimeException("NLU")
    }

}
